docs/man/Module.3
.TH "Module" 3 "30 Apr 2004" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Module \- Base class for all InspIRCd modules This class is the base class for InspIRCd modules. 

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <modules.h>\fP
.PP
Inherits \fBclassbase\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBModule\fP ()"
.br
.RI "\fIDefault constructor creates a module class.\fP"
.ti -1c
.RI "virtual \fB~Module\fP ()"
.br
.RI "\fIDefault destructor destroys a module class.\fP"
.ti -1c
.RI "virtual \fBVersion\fP \fBGetVersion\fP ()"
.br
.RI "\fIReturns the version number of a Module.\fP"
.ti -1c
.RI "virtual void \fBOnUserConnect\fP (\fBuserrec\fP *user)"
.br
.RI "\fICalled when a user connects.\fP"
.ti -1c
.RI "virtual void \fBOnUserQuit\fP (\fBuserrec\fP *user)"
.br
.RI "\fICalled when a user quits.\fP"
.ti -1c
.RI "virtual void \fBOnUserJoin\fP (\fBuserrec\fP *user, \fBchanrec\fP *channel)"
.br
.RI "\fICalled when a user joins a channel.\fP"
.ti -1c
.RI "virtual void \fBOnUserPart\fP (\fBuserrec\fP *user, \fBchanrec\fP *channel)"
.br
.RI "\fICalled when a user parts a channel.\fP"
.ti -1c
.RI "virtual void \fBOnPacketTransmit\fP (char *p)"
.br
.RI "\fICalled before a packet is transmitted across the irc network between two irc servers.\fP"
.ti -1c
.RI "virtual void \fBOnPacketReceive\fP (char *p)"
.br
.RI "\fICalled after a packet is received from another irc server.\fP"
.ti -1c
.RI "virtual void \fBOnRehash\fP ()"
.br
.RI "\fICalled on rehash.\fP"
.ti -1c
.RI "virtual void \fBOnServerRaw\fP (std::string &raw, bool inbound, \fBuserrec\fP *user)"
.br
.RI "\fICalled when a raw command is transmitted or received.\fP"
.ti -1c
.RI "virtual bool \fBOnExtendedMode\fP (\fBuserrec\fP *user, void *target, char modechar, int type, bool mode_on, \fBstring_list\fP &params)"
.br
.RI "\fICalled whenever an extended mode is to be processed.\fP"
.ti -1c
.RI "virtual int \fBOnUserPreJoin\fP (\fBuserrec\fP *user, \fBchanrec\fP *chan, const char *cname)"
.br
.RI "\fICalled whenever a user is about to join a channel, before any processing is done.\fP"
.ti -1c
.RI "virtual void \fBOnOper\fP (\fBuserrec\fP *user)"
.br
.RI "\fICalled whenever a user opers locally.\fP"
.ti -1c
.RI "virtual void \fBOnInfo\fP (\fBuserrec\fP *user)"
.br
.RI "\fICalled whenever a user types /INFO.\fP"
.ti -1c
.RI "virtual void \fBOnWhois\fP (\fBuserrec\fP *source, \fBuserrec\fP *dest)"
.br
.RI "\fICalled whenever a /WHOIS is performed on a local user.\fP"
.ti -1c
.RI "virtual int \fBOnUserPreMessage\fP (\fBuserrec\fP *user, void *dest, int target_type, std::string text)"
.br
.RI "\fICalled whenever a user is about to PRIVMSG A user or a channel, before any processing is done.\fP"
.ti -1c
.RI "virtual int \fBOnUserPreNotice\fP (\fBuserrec\fP *user, void *dest, int target_type, std::string text)"
.br
.RI "\fICalled whenever a user is about to NOTICE A user or a channel, before any processing is done.\fP"
.ti -1c
.RI "virtual int \fBOnUserPreNick\fP (\fBuserrec\fP *user, std::string newnick)"
.br
.RI "\fICalled before any nickchange, local or remote.\fP"
.in -1c
.SH "Detailed Description"
.PP 
Base class for all InspIRCd modules This class is the base class for InspIRCd modules.

All modules must inherit from this class, its methods will be called when irc server events occur. class inherited from module must be instantiated by the \fBModuleFactory\fP class (see relevent section) for the plugin to be initialised. 
.PP
Definition at line 92 of file modules.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Module::Module ()"
.PP
Default constructor creates a module class.Definition at line 111 of file modules.cpp.
.PP
.nf
111 { }
.fi
.SS "Module::~Module ()\fC [virtual]\fP"
.PP
Default destructor destroys a module class.Definition at line 112 of file modules.cpp.
.PP
.nf
112 { }
.fi
.SH "Member Function Documentation"
.PP 
.SS "\fBVersion\fP Module::GetVersion ()\fC [virtual]\fP"
.PP
Returns the version number of a Module.The method should return a \fBVersion\fP object with its version information assigned via \fBVersion::Version\fPDefinition at line 123 of file modules.cpp.
.PP
.nf
123 { return Version(1,0,0,0); }
.fi
.SS "bool Module::OnExtendedMode (\fBuserrec\fP * user, void * target, char modechar, int type, bool mode_on, \fBstring_list\fP & params)\fC [virtual]\fP"
.PP
Called whenever an extended mode is to be processed.The type parameter is MT_SERVER, MT_CLIENT or MT_CHANNEL, dependent on where the mode is being changed. mode_on is set when the mode is being set, in which case params contains a list of parameters for the mode as strings. If mode_on is false, the mode is being removed, and parameters may contain the parameters for the mode, dependent on wether they were defined when a mode handler was set up with \fBServer::AddExtendedMode\fP If the mode is a channel mode, target is a chanrec*, and if it is a user mode, target is a userrec*. You must cast this value yourself to make use of it.Definition at line 122 of file modules.cpp.
.PP
.nf
122 { return false; }
.fi
.SS "void Module::OnInfo (\fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Called whenever a user types /INFO.The userrec will contain the information of the user who typed the command. Modules may use this method to output their own credits in /INFO (which is the ircd's version of an about box). It is purposefully not possible to modify any info that has already been output, or halt the list. You must write a 371 numeric to the user, containing your info in the following format:
.PP
:information hereDefinition at line 125 of file modules.cpp.
.PP
.nf
125 { };
.fi
.SS "void Module::OnOper (\fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Called whenever a user opers locally.The userrec will contain the oper mode 'o' as this function is called after any modifications are made to the user's structure by the core.Definition at line 124 of file modules.cpp.
.PP
.nf
124 { };
.fi
.SS "void Module::OnPacketReceive (char * p)\fC [virtual]\fP"
.PP
Called after a packet is received from another irc server.The packet is represented as a char*, as it should be regarded as a buffer, and not a string. This allows you to easily represent it in the correct ways to implement encryption, compression, digital signatures and anything else you may want to add. This should be regarded as a pre-processor and will be called immediately after the packet is received but before any other operations with the core of the ircd.Definition at line 118 of file modules.cpp.
.PP
.nf
118 { }
.fi
.SS "void Module::OnPacketTransmit (char * p)\fC [virtual]\fP"
.PP
Called before a packet is transmitted across the irc network between two irc servers.The packet is represented as a char*, as it should be regarded as a buffer, and not a string. This allows you to easily represent it in the correct ways to implement encryption, compression, digital signatures and anything else you may want to add. This should be regarded as a pre-processor and will be called before ANY other operations within the ircd core program.Definition at line 117 of file modules.cpp.
.PP
.nf
117 { }
.fi
.SS "void Module::OnRehash ()\fC [virtual]\fP"
.PP
Called on rehash.This method is called prior to a /REHASH or when a SIGHUP is received from the operating system. You should use it to reload any files so that your module keeps in step with the rest of the application.Definition at line 119 of file modules.cpp.
.PP
.nf
119 { }
.fi
.SS "void Module::OnServerRaw (std::string & raw, bool inbound, \fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Called when a raw command is transmitted or received.This method is the lowest level of handler available to a module. It will be called with raw data which is passing through a connected socket. If you wish, you may munge this data by changing the string parameter 'raw'. If you do this, after your function exits it will immediately be cut down to 510 characters plus a carriage return and linefeed. For INBOUND messages only (where inbound is set to true) the value of user will be the userrec of the connection sending the data. This is not possible for outbound data because the data may be being routed to multiple targets.Definition at line 120 of file modules.cpp.
.PP
.nf
120 { }
.fi
.SS "void Module::OnUserConnect (\fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Called when a user connects.The details of the connecting user are available to you in the parameter userrec *userDefinition at line 113 of file modules.cpp.
.PP
.nf
113 { }
.fi
.SS "void Module::OnUserJoin (\fBuserrec\fP * user, \fBchanrec\fP * channel)\fC [virtual]\fP"
.PP
Called when a user joins a channel.The details of the joining user are available to you in the parameter userrec *user, and the details of the channel they have joined is available in the variable chanrec *channelDefinition at line 115 of file modules.cpp.
.PP
.nf
115 { }
.fi
.SS "void Module::OnUserPart (\fBuserrec\fP * user, \fBchanrec\fP * channel)\fC [virtual]\fP"
.PP
Called when a user parts a channel.The details of the leaving user are available to you in the parameter userrec *user, and the details of the channel they have left is available in the variable chanrec *channelDefinition at line 116 of file modules.cpp.
.PP
.nf
116 { }
.fi
.SS "int Module::OnUserPreJoin (\fBuserrec\fP * user, \fBchanrec\fP * chan, const char * cname)\fC [virtual]\fP"
.PP
Called whenever a user is about to join a channel, before any processing is done.Returning any nonzero value from this function stops the process immediately, causing no output to be sent to the user by the core. If you do this you must produce your own numerics, notices etc. This is useful for modules which may want to mimic +b, +k, +l etc.
.PP
IMPORTANT NOTE!
.PP
If the user joins a NEW channel which does not exist yet, OnUserPreJoin will be called BEFORE the channel record is created. This will cause chanrec* chan to be NULL. There is very little you can do in form of processing on the actual channel record at this point, however the channel NAME will still be passed in char* cname, so that you could for example implement a channel blacklist or whitelist, etc.Definition at line 121 of file modules.cpp.
.PP
.nf
121 { return 0; }
.fi
.SS "int Module::OnUserPreMessage (\fBuserrec\fP * user, void * dest, int target_type, std::string text)\fC [virtual]\fP"
.PP
Called whenever a user is about to PRIVMSG A user or a channel, before any processing is done.Returning any nonzero value from this function stops the process immediately, causing no output to be sent to the user by the core. If you do this you must produce your own numerics, notices etc. This is useful for modules which may want to filter or redirect messages. target_type can be one of TYPE_USER or TYPE_CHANNEL. If the target_type value is a user, you must cast dest to a userrec* otherwise you must cast it to a chanrec*, this is the details of where the message is destined to be sent.Definition at line 127 of file modules.cpp.
.PP
.nf
127 { return 0; };
.fi
.SS "int Module::OnUserPreNick (\fBuserrec\fP * user, std::string newnick)\fC [virtual]\fP"
.PP
Called before any nickchange, local or remote.This can be used to implement Q-lines etc. Please note that although you can see remote nickchanges through this function, you should NOT make any changes to the userrec if the user is a remote user as this may cause a desnyc. check user->server before taking any action (including returning nonzero from the method). If your method returns nonzero, the nickchange is silently forbidden, and it is down to your module to generate some meaninful output.Definition at line 129 of file modules.cpp.
.PP
.nf
129 { return 0; };
.fi
.SS "int Module::OnUserPreNotice (\fBuserrec\fP * user, void * dest, int target_type, std::string text)\fC [virtual]\fP"
.PP
Called whenever a user is about to NOTICE A user or a channel, before any processing is done.Returning any nonzero value from this function stops the process immediately, causing no output to be sent to the user by the core. If you do this you must produce your own numerics, notices etc. This is useful for modules which may want to filter or redirect messages. target_type can be one of TYPE_USER or TYPE_CHANNEL. If the target_type value is a user, you must cast dest to a userrec* otherwise you must cast it to a chanrec*, this is the details of where the message is destined to be sent.Definition at line 128 of file modules.cpp.
.PP
.nf
128 { return 0; };
.fi
.SS "void Module::OnUserQuit (\fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Called when a user quits.The details of the exiting user are available to you in the parameter userrec *userDefinition at line 114 of file modules.cpp.
.PP
.nf
114 { }
.fi
.SS "void Module::OnWhois (\fBuserrec\fP * source, \fBuserrec\fP * dest)\fC [virtual]\fP"
.PP
Called whenever a /WHOIS is performed on a local user.The source parameter contains the details of the user who issued the WHOIS command, and the dest parameter contains the information of the user they are whoising.Definition at line 126 of file modules.cpp.
.PP
.nf
126 { };
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for InspIRCd from the source code.


docs/ChangeLog
1.0 Final

    * Optimized WriteServ - why use %s!%s@%s when we have the more optimal userrec::GetFullHost() ?
    * Added new event into m_spanningtree, event id send_metadata, which allows arbitary posting of metadata outside of the netburst, see line 529 of m_ssl_gnutls for an example of how to use it.
    * Added check that forbids configuration as root >:-)
    * Major updates to CBANs...like they're reasonably functional
    * Added support for TYPE_OTHER metadata, where void* target will be NULL, e.g. its not bound to a user or a nick
    * Fix to send 403 on invalid channel name
    * Moved Write() call so that it can send the NICK message properly and force the client to change
    * Allow ulines to still notice the channel regardless of +T
    * Created OnPostOper that can be gauranteed to be called *after* opering (by ALL modules) is done
    * "Oper of unknown type" -> "313 %s %s :is an IRC operator"
    * Auto-assume "+" on the start of opermode block if not included
    * Removed example commands that never actually existed
    * OpenSSL detection and cert generation in ./configure
    * Changed length of password field to 64, giving us a maximum password length of 63 plus null terminator
    * Fixed uninitialized ExtMode::list
    * Added some extra debug
    * Make m_services use InsertMode (are we done yet?)
    * Changed On005Numeric to use InsertMode (we left one out?)
    * Deleted the semicolon that had half the dev team wondering...
    * Incorporated Craig's friendlier SQL class (documentation to come later)
    * m_sqlauth: Fix small memory leaks in some cases (bad password, bad query, lost mysql server connection etc..)
    * Correctly initialize userrec::password (should fix a few m_sql issues)
    * Add m_conn_waitpong, require clients to reply with a correct PONG to a server PING before they can connect
    * Remove unused includes and rename classes from ModuleSSL to ModuleSSLGnuTLS so we can be consistent across multiple SSL modules
    * Added warnings to opers when session limits are exceeded
    * Improved IP handling. Now uses in_addr to store client ip, not char[16]!, Added global and local session limits
    * Problems removing +t due to wrong bit-value being used in comparison
    * Altered so that if running in -debug mode, will not erase its .so files from /tmp to allow a clean backtrace
    * Removed unneccessary strlcpy
    * Removed a load of unneeded includes (poll.h etc)
    * Added <ident:timeout> example so that people remember to set this when using m_ident.so
    * Default ident timeout to 1 sec if not defined (a bit low, but better than 0!)
    * Updated m_ident to use inspsocket, should be more reliable
    * Added missing modules (newer ones)
    * Removed excessive debug for something thats now fixed
    * Merged Andy Church's isnick() and isident() patches - faster, more efficient, and doesnt accidentally allow "-" as the first character of a nickname, Also adapted to iterate over a char* pointer rather than using unsigned int for the for() loops
    * Added m_opersha256 module created by Special
    * Add module for anti-autorejoin channel mode +J
    * Changes to INVEX and EXCEPTS 005 tokens
    * Changed +l to drop the mode if the limit is same as current., Optimized check on numeric chars
    * Remove last relic of last goto in inspircd (an unused label)
    * Fixed to not have ohv in CHANMODES type A,, Added proper MAXMODES output in 005 instead of hard coded '13'
    * One goto is now no gotos.
    * Change m_redirect, m_joinflood and m_messageflood to put their modes in the correct section and to use InsertMode()
    * Add InsertMode function to helperfuncs.(cpp|h) for easily adding modes to the CHANMODES section of the 005 numeric
    * Added patch from Andy Church re spaces in SETNAME
    * Fixed line endings when skipping ssl cert generation
    * Documented some functions that were missing comments
    * Properly check that existing key resides in conf/ and not current dir
    * Added PRIORITY_BEFORE and PRIORITY_AFTER (see src/modules/m_hostchange.cpp for how it works, function Prioritize())
    * Added -modupdate to regenerate modules makefile, Added SSL option, as well as a 'Build Certificate' section (If avaliable)
    * Adding the +e and +I modules m_banexception and m_inviteexception, also adding their shared utility header
    * Added trick to allow local server to unset -r only in one specific condition (the nickchange)
    * Added joinflood +j
    * Made bans check both real and fake host
    * Patched to allow FTOPIC from users (so that chanserv can update the settime and the topic still come from chanserv)
    * Added <link:hidden> to docs
    * Optional IP-Hiding for servers
    * Use libgnutls-config --libs to configure linker options
    * Make relative paths in the config file tags be relative to the insp config directory
    * Change 974 numeric to 490 to avoid collision with Insp's failed to load module error
    * Make m_sslmodes messages appear in the channel window
    * Updated with new docs for maxtargets
    * Updated 005 to show changeable MAXTARGETS
    * Allowed for configuration of maxtargets in <options:maxtargets>
    * Wrong nicklen is shown in the 005 numeric, fixed (noticed by ThaPrince)
    * Initial commit of m_sslmodes, supplying chanmode +z
    * Allow for <bind> tags with no or empty type, as well as 'clients' ones
    * Set vendor and static module flags...even though the booms from trying to unload it would be fun sad.gif
    * Make errors loading certificates DEFAULT level
    * Inital commit of the first SSL module \o/ \o/ \o/ Also Ommeh's first official svn commit \o/
    * Added MakeWildHost to userrec (*!*@dhost), Added ban stuff to messageflood
    * Fixes to server kick (i hope)
    * And yae, the compiler sayeth to brain, fix thine warning. and thou warning art fixeth.
    * Made '*' ban-specifier work

src/python/module.py
#
# InspIRCd -- Internet Relay Chat Daemon
#
#   Copyright (C) 2026 Sadie Powell <sadie@sadiepowell.dev>
#
# This file is part of InspIRCd.  InspIRCd is free software: you can
# redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


import pathlib

from . import common

DIR         = common.ROOT / "modules"
CONTRIB_DIR = DIR         / "contrib"
EXTRA_DIR   = DIR         / "extra"


# Converts a module name to a file name.
def file_name(module):
    return pathlib.Path(module).with_suffix(".cpp").name


# Convert a module file name or path to a short name.
def short_name(module):
    return pathlib.Path(module).stem


# Compares two module versions and returns -1 if older, 0 if equal, or 1 if newer.
def version_compare(lhs, rhs):
    for lhs, rhs in zip(lhs, rhs):
        if lhs is None or rhs is None:
            break
        elif lhs < rhs:
            return -1
        elif lhs > rhs:
            return 1
    return 0


# Checks whether a module version matches a requirement range.
def version_in_range(version, requirement):
    if not version:
        return requirement

    min_requirement, max_requirement = requirement
    if version_compare(min_requirement, version) > 0:
        return False
    elif version_compare(version, max_requirement) > 0:
        return False

    return True


# Converts a module version string to a tuple of three elements.
def version_parse(version):
    if not version:
        return [None, None, None]

    segments = [int(s) for s in version.split(".") if s.isdigit()]
    return tuple((segments + [None, None, None])[:3])


# Converts a tuple of three elements to a module version string.
def version_string(segments):
    return ".".join(str(segment) for segment in segments if segment is not None)


# Converts a two tuples of three elements to a module version range.
def version_range_string(range):
    return "-".join(version_string(version) for version in set(range))
